enhancements/olm-curated-catalog.md

```diff
@@ -0,0 +1,353 @@
+
+## Proposal
+
+The basic approach will be to:
+
+- Generate catalog index container images that contain a subset of existing catalog container images
+- Provide the ability to merge catalogs together, to allow periodic mirroring of catalogs that are updated over time.
+- When filtering by version:
+  - Update any broken graph links with a subset of the new, sparse graph
+- When filtering by os/arch:
+  - Update the CSVs in the catalog index to have updated `relatedImages` references to the sparse manifest indexes 
+  - Update the CSVs in the catalog index to have updated `operatorframework.io/arch` and `operatorframework.io/os` labels.
+  - This requires the multi-architecture images to be mirrored to a repository with the appropriate sparse manifest index created.
+
+
+### Generate a curated catalog
+
+`opm index build` will be created with flags:
+
+```sh
+$ opm index build \ 
+  --from-index  "quay.io/openshift/community-operators-catalog:latest"
+  --package etcd \
+  --filter-by-channel "beta|stable"\
+  --default-channel "stable" \
+  --filter-by-version=">=0.9.2" \
+  --filter-by-os="linux\/amd64|linux\/ppc64le" \
+  --tag="disconnected-registry:5000/catalogs/curated-catalog:latest"
+  --merge-index
+```
+
+It will have the following additional flags:
+- `--from-index` - the index image to use as the source to retrieve operators from.
+- `--package` - the name of an operator package to include.
+- `--filter-by-channel` - a regular expression of one or more channels to include from the package.
+- `--default-channel` - the default channel to set for the package.
+- `--filter-by-version` - a semver range identifying a subset of the version graph (DAG) to include from the package.
+- `--filter-by-os` - a regular expression of operating system and architectures to support.
+- `--merge-index` - the index image to merge the filtered operator versions into.
+
+The following flags are inherited from the `opm index add` command:
+```
+  -i, --binary-image opm        container image for on-image opm command
+  -c, --container-tool string   tool to interact with container images (save, build, etc.). One of: [docker, podman] (default "podman")
+      --generate                if enabled, just creates the dockerfile and saves it to local disk
+  -h, --help                    help for add
+  -d, --out-dockerfile string   if generating the dockerfile, this flag is used to (optionally) specify a dockerfile name
+      --permissive              allow registry load errors
+  -t, --tag string              custom tag for container image being built
+```
+
+`opm index build` will be enhanced to:
+
+- Retrieve the source Catalog Index Image
+- Extract the Catalog Index Image that match the following filters into a separate directory representing an Operator Bundle (CRDs, CSVs, Metadata).
+  - Package Name (the operator name)
+  - Channel names
+  - Version range to include (semver range)
+- If the `filter-by-version` is specified:
```

Comment:
I also wonder how this is aggregated per package? Unless we are talking about having unique indexes per package that are all zipped up one at a time per this command?

Reply:
> I also wonder how this is aggregated per package? Unless we are talking about having unique indexes per package that are all zipped up one at a time per this command?

Agree. I think it will be difficult to specify the contents of an index on the command line. There probably needs to be some kind of manifest that gets passed in, something like a list of: (package, channel, optional version expression)

Reply:
Yes, I actually started with that concept, but thought it got really complicated really fast.  

One could easily model the entire set of packages and their graphs using some sort of grammar.  I wanted to start simple like this to make sure everyone was on the same page.